### PR TITLE
[CDU] Allow SimBrief user IDs instead of usernames

### PR DESCRIPTION
**Pull request by pareil6**
## Summary of Changes
Allows user to insert a SimBrief user ID instead of username.

## References
N/A

## Additional context
N/A

Discord username (if different from GitHub): pareil6#0990

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by ZigTag — APPROVED** [review]
Code LGTM

**Comment by pareil6** [comment]
CDU font doesn't seem to support underscores, so currently displaying them as spaces instead of a rectangle - unless there's a trick to get an underscore to show up?

Also might be worth making this obvious on the options menu - open to suggestions for the cleanest way to do it.

**Review by ZigTag — CHANGES_REQUESTED** [review]
LGTM but add Changelog

**Review by ZigTag — APPROVED** [review]
Code LGTM

**Comment by Benjozork** [comment]
cc. @viniciusfont

**Comment by pareil6** [comment]
Will need a QA tester with a Simbrief username with an underscore - I can't check functionality beyond it displaying correctly since mine doesn't.

**Comment by ImenesFBW** [comment]
@tareksabet your time to shine, underscore man!

**Comment by pareil6** [comment]
Have now woken up - more obvious than slashes, spaces also aren't legal in SB usernames, so use spaces instead of slashes (which satisfies my obsessive compulsive streak because we can display them the same as they're typed).

**Comment by wpine215** [comment]
> Have now woken up - more obvious than slashes, spaces also aren't legal in SB usernames, so use spaces instead of slashes (which satisfies my obsessive compulsive streak because we can display them the same as they're typed).

This sounds like the best solution. I can push a quick update to the font later today to add support for displaying underscores.

**Comment by tareksabet** [comment]
> @tareksabet your time to shine, underscore man!

##too late Imenes I already changed my username

**Comment by viniciusfont** [comment]
I did not tested, but the code looks good! 👍

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ class CDU_OPTIONS_SIMBRIEF {
     static ShowPage(mcdu) {
         mcdu.clearDisplay();
 
-        const simbriefUsername = NXDataStore.get("CONFIG_SIMBRIEF_USERNAME", "");
+        const simbriefUsername = NXDataStore.get("CONFIG_SIMBRIEF_USERNAME", "").replace('_', ' ');
 
         const simbriefUsernameString = simbriefUsername != "" ? simbriefUsername : "[ ]";
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js:7]:
```suggestion
        const simbriefUsername = NXDataStore.get("CONFIG_SIMBRIEF_USERNAME", "");

        const simbriefUsernameString = simbriefUsername ? simbriefUsername..replace('_', ' ') : "[ ]";
```

**Review by MisterChocker — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js:7]: or use `-` instead

**Comment by STEELBADGE** [comment]
We should use the Simbrief ID instead of the username this removes the need for any hacks to account for special characters in the username.

**Review by devsnek — CHANGES_REQUESTED** [review]
`s.replace(n, r)` only replaces the first occurrence of `n`. You should do `s.replace(/n/g, r)` instead.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js (hunk 1 above)
Comment (pareil6) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js:7]: Is it not cleaner to store the correct username (ie. with underscores) in the DataStore, and deal with rewriting it when saving/displaying it? This way around, anywhere using  NXDataStore.get("CONFIG_SIMBRIEF_USERNAME") doesn't need to know what character we're substituting for underscores, we can keep that here instead.

Also, is '-' allowed in Simbrief usernames? I'd assumed it was legal which was why I didn't use it, but happy to switch it around if it's not.

**Comment by pareil6** [comment]
> We should use the Simbrief ID instead of the username this removes the need for any hacks to account for special characters in the username.

I'd agree with this to be honest, especially given usernames can change. It's relatively simple to support both (and allow the user to enter one or the other), which would avoid this bodge...

File: A32NX/html_ui/Pages/A32NX_Utils/SimBriefApi.js [hunk 2]
```diff
@@ -1,10 +1,15 @@
 class SimBriefApi {
-    static getSimBriefOfp(username) {
-        if (!username) {
-            throw ("No SimBrief username provided");
+    static getSimBriefOfp(username, userId) {
+        var apiUrl;
```
Comment (viniciusfont) [inline A32NX/html_ui/Pages/A32NX_Utils/SimBriefApi.js:3]: Use let instead of var for code consistency

**Comment by pareil6** [comment]
Added option to provide a user ID instead of username. Name/ID are mutually exclusive, so entering one will clear the other.

I've left the space/underscore substitution in on top of this, happy to remove it if everyone would prefer though.

File: A32NX/html_ui/Pages/A32NX_Utils/SimBriefApi.js (hunk 2 above)
Comment (pareil6) [reply A32NX/html_ui/Pages/A32NX_Utils/SimBriefApi.js]: Non-JS dev figuring this out as I go along 😛 Thanks, fixed.

Comment (viniciusfont) [reply A32NX/html_ui/Pages/A32NX_Utils/SimBriefApi.js]: JS is confusing even when declaring a simple variable 🤣

**Review by Benjozork — APPROVED** [review]
LGGTM

**Comment by MisterChocker** [comment]
> Added option to provide a user ID instead of username. Name/ID are mutually exclusive, so entering one will clear the other.
> 
> I've left the space/underscore substitution in on top of this, happy to remove it if everyone would prefer though.

why not use one field only? like the one existing?

**Comment by Road-to-Hell9** [comment]
Category: QA Tester
Name: Road_to_Hell9 #1111
Date of testing: 19/12/2020
Version of the sim: 19/12/2020, most recent
PR Tested: #2496
Object of test: [CDU] Allow SimBrief user IDs instead of usernames

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1.Entered the ID on the settings
2. Filed a flight plan
3. Got the flight plan form the Init page

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: It works, but you could add the ZFW, Fuel and the CG, if the CG can be requested form the sim
Screenshot (118)
Screenshot (116)

**Comment by MisterChocker** [comment]
> Conclusions of the testing: It works, but you could add the ZFW, Fuel and the CG, if the CG can be requested form the sim in the init B page

this could indeed be done

**Comment by Benjozork** [comment]
> but you could add the ZFW, Fuel and the CG, if the CG can be requested form the sim in the init B page

Unrelated to the PR

**Comment by pareil6** [comment]
> > Added option to provide a user ID instead of username. Name/ID are mutually exclusive, so entering one will clear the other.
> > I've left the space/underscore substitution in on top of this, happy to remove it if everyone would prefer though.
> 
> why not use one field only? like the one existing?

I figure everyone will know their username, but most will need to go to SimBrief to look up their user ID - so it felt more user friendly to leave the username option there by default, but with the fallback of the ID being something that every SimBrief user will be able to enter.

**Comment by MisterChocker** [comment]
> I figure everyone will know their username, but most will need to go to SimBrief to look up their user ID - so it felt more user friendly to leave the username option there, but with the fallback of the ID being something that every SimBrief user will be able to enter.

I didn't meant to scrub the username option, the idea was to have one field that accepts both

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js]: this doesn't modify the stored data, just the displayed one

Comment (pareil6) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/OPTIONS/A320_Neo_CDU_OPTIONS_AOC_SIMBRIEF.js]: The modification happens further down when the user enters their username though

**Comment by pareil6** [comment]
> > I figure everyone will know their username, but most will need to go to SimBrief to look up their user ID - so it felt more user friendly to leave the username option there, but with the fallback of the ID being something that every SimBrief user will be able to enter.
> 
> I didn't meant to scrub the username option, the idea was to have one field that accepts both

Ah OK - I think all-numerical usernames are legal though so it's not necessarily obvious which one a user means, which is why I thought it was clearer to separate the options out.

**Comment by valenzz2004** [comment]
Category: QA Tester
Name: Valenzz2004#7802
Date of testing: 19/12/2020
Version of the sim: 1.11.6.0
PR Tested:#2496
Object of test: [CDU] Allow SimBrief user IDs instead of usernames

Testing technique: uplink flight from simbrief using ID only
Steps you've made to test the Pull Request:

1. Loaded in to Bahrain OBBI
2. removed previous username and added pilot id
3.pressed init request and loaded in plan to OMDB
4. loaded fuel and not payload (known issue)

Flight notes (If required):

Issues (If Any): 
Overall Rating: Good 
Conclusions of the testing: should be a great replacement for simbrief username. 100% functional.
Screenshots (if any):
simbrief test
Screenshot 2020-12-19 174641
simbrief qa

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -5,7 +5,7 @@
 <!--  1. [Changed Area] Title of changes - @github username (Name)  -->
 
 ## 0.6.0
-1.
+1. [CDU] Allow SimBrief user IDs as well as usernames
```
Comment (pepperoni505) [inline .github/CHANGELOG.md:8]: Add your username

Comment (pareil6) [reply .github/CHANGELOG.md]: Good spot, fixed.

**Comment by Awemeter** [comment]
Category: QA Tester
Name: Awemeter#7919
Date of testing: 12/19/2020
Version of the sim: 1.11.7.0
PR Tested: #2496 
Object of test: Sim brief integration 
Testing technique: Tried out
Steps you've made to test the Pull Request:
1. Spawned in sim
2. Went to options in mcdu and clicked simbreif
3.  Added user-id
4. Requested int and data to see if it still worked. 
5. 

Flight notes (If required):

Issues (If Any): N/A
Overall Rating: Good
Conclusions of the testing: No issues works like a dream. Keep up the good work pareil6

**Pull request merged**